atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStore.java

```diff
@@ -0,0 +1,166 @@
+        shuffleNodes();
+        ListenableFuture<Optional<byte[]>> read = get(cell);
+        if (!Arrays.equals(AtlasFutures.getDone(read).get(), value)) {
+            throw new CheckAndSetException("not happening");
+        }
+
+        writeToQuorum(cell, value);
+    }
+
+    @Override
+    public ListenableFuture<Optional<byte[]>> get(Cell cell) {
+        shuffleNodes();
+        Optional<BytesAndTimestamp> latest = IntStream.range(0, quorum)
+                .mapToObj(nodes::get)
+                .map(node -> node.get(cell))
+                .filter(Objects::nonNull)
+                .max(Comparator.comparing(BytesAndTimestamp::timestamp));
+        if (latest.isEmpty()) {
+            return Futures.immediateFuture(Optional.empty());
+        }
+        BytesAndTimestamp latestPresent = latest.get();
+        for (int i = 0; i < quorum; i++) {
+            nodes.get(i).put(cell, latestPresent);
+        }
+        return Futures.immediateFuture(latest.map(BytesAndTimestamp::bytes));
+    }
+
+    @Override
+    public ListenableFuture<Map<Cell, byte[]>> getMultiple(Iterable<Cell> cells) {
+        // too much effort to implement general case since it won't be used
+        Cell onlyCell = Iterables.getOnlyElement(cells);
+        return Futures.transform(
+                get(onlyCell),
+                maybeValue -> maybeValue
+                        .map(value -> ImmutableMap.of(onlyCell, value))
+                        .orElseGet(ImmutableMap::of),
+                MoreExecutors.directExecutor());
+    }
+
+    @Override
+    public void put(Cell cell, byte[] value) {
+        shuffleNodes();
+        for (int i = 0; i < quorum; i++) {
+            nodes.get(i).put(cell, ImmutableBytesAndTimestamp.of(value, KvsConsensusForgettingStore.PUT_TIMESTAMP));
+            maybeFail();
+        }
+    }
+
+    @Override
+    public void put(Map<Cell, byte[]> values) {
+        values.forEach(this::put);
+    }
+
+    private void shuffleNodes() {
+        Collections.shuffle(nodes);
+    }
+
+    private void maybeFail() {
+        if (random.nextDouble() <= probabilityOfFailure) {
+            throw new SafeRuntimeException("Ohno!");
+        }
+    }
+
+    private void writeToQuorum(Cell cell, byte[] value) {
+        maybeFail();
+        long ts = timestamps.getAndIncrement();
+        for (int i = 0; i < quorum; i++) {
+            nodes.get(i).put(cell, ImmutableBytesAndTimestamp.of(value, ts));
+            maybeFail();
+        }
+    }
+
+    @Value.Immutable
+    interface BytesAndTimestamp extends Comparable<BytesAndTimestamp> {
+        @Value.Parameter
+        byte[] bytes();
+
+        @Value.Parameter
+        long timestamp();
+
+        @Override
+        default int compareTo(BytesAndTimestamp other) {
+            if (other == null) {
+                return 1;
+            }
+            return Comparator.comparing(BytesAndTimestamp::timestamp).compare(this, other);
+        }
```

Comment:
Not sure I understand why this is needed: it seems you use a specific Comparator when finding the latest value anyway.

Reply:
Just for the method below